src/agent0/hyperfuzz/system_fuzz/invariant_checks.py

```diff
@@ -202,6 +205,52 @@ class InvariantCheckResults(NamedTuple):
     log_level: int | None
 
 
+def _check_price_spike(interface: HyperdriveReadInterface, pool_state: PoolState) -> InvariantCheckResults:
+    # Warn if a trade causes a large rate spike; which is a precursor to an attack
+    failed = False
+    exception_message: str | None = None
+    exception_data: dict[str, Any] = {}
+    log_level = None
+    delta_rate_epsilon = interface.pool_config.circuit_breaker_delta
+
+    # TODO we hack in a stateful variable into the interface here to check
+    # between subsequent calls.
+    # Initial call, we look to see if the attribute exists
+    previous_pool_state: PoolState | None = getattr(interface, "_price_spike_previous_pool_state", None)
+    # Always set the new state here
+    setattr(interface, "_price_spike_previous_pool_state", pool_state)
+
+    if previous_pool_state is None:
+        # Skip this check on initial call, not a failure
+        return InvariantCheckResults(
+            failed=False, exception_message=exception_message, exception_data=exception_data, log_level=log_level
+        )
+
+    # The checkpoint weighted spot price is updated every checkpoint and every trade.
+    previous_weighted_spot_apr = interface.calc_rate_given_fixed_price(
+        previous_pool_state.checkpoint.weighted_spot_price,
+        FixedPoint(previous_pool_state.pool_config.position_duration),
+    )
+    current_weighted_spot_apr = interface.calc_rate_given_fixed_price(
+        pool_state.checkpoint.weighted_spot_price,
+        FixedPoint(pool_state.pool_config.position_duration),
+    )
+
+    if abs(current_weighted_spot_apr - previous_weighted_spot_apr) >= delta_rate_epsilon:
```

Comment:
It seems to me that this warning is only going to let us know when the attack is possible. Could we instead do this update when the spot price is greater than the previous weighted spot price? This would tell us that there may be a problem as soon as the trade occurs.

Reply:
greater than? would this not happen every long?